In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Cargar el dataset CIFAR-10
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

# Normalización de imágenes (Las imágenes en el dataset CIFAR-10 tienen valores de píxeles que van de 0 a 255.)
x_train, x_test = x_train / 255.0, x_test / 255.0

# Data Augmentation (Se utiliza Data Augmentation para generar nuevas variaciones de las imágenes del conjunto de entrenamiento.)
datagen = ImageDataGenerator(
    rotation_range=15,
    width_shift_range=0.1,
    height_shift_range=0.1,
    horizontal_flip=True
)
datagen.fit(x_train)

# Definir el modelo CNN ( Modelo CNN con 3 capas conv, max pooling, flatten, 64 neuronas densas y salida softmax para 10 clases)
model = models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax') # 10 clases de salida con softmax para clasificación
])

# Compilar el modelo
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy', #Para este problema de clasificación multiclase, la función de pérdida más adecuada es sparse_categorical_crossentropy, 
                                                      #ya que nuestras etiquetas están en formato entero, lo cual se ajusta perfectamente a este tipo de tareas.
              metrics=['accuracy'])

# Entrenamiento con Data Augmentation
history = model.fit(datagen.flow(x_train, y_train, batch_size=64),
                    epochs=10, 
                    validation_data=(x_test, y_test))

# Evaluar el modelo
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)

print(f"Test accuracy: {test_acc}")


170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 26s 0us/step


C:\Users\diego\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10


C:\Users\diego\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


782/782 ━━━━━━━━━━━━━━━━━━━━ 49s 59ms/step - accuracy: 0.3141 - loss: 1.8437 - val_accuracy: 0.5092 - val_loss: 1.3511
Epoch 2/10
782/782 ━━━━━━━━━━━━━━━━━━━━ 48s 61ms/step - accuracy: 0.4858 - loss: 1.4221 - val_accuracy: 0.5551 - val_loss: 1.2402
Epoch 3/10
782/782 ━━━━━━━━━━━━━━━━━━━━ 60s 76ms/step - accuracy: 0.5455 - loss: 1.2772 - val_accuracy: 0.6200 - val_loss: 1.0807
Epoch 4/10
782/782 ━━━━━━━━━━━━━━━━━━━━ 61s 78ms/step - accuracy: 0.5766 - loss: 1.1903 - val_accuracy: 0.5997 - val_loss: 1.1287
Epoch 5/10
782/782 ━━━━━━━━━━━━━━━━━━━━ 45s 57ms/step - accuracy: 0.6009 - loss: 1.1326 - val_accuracy: 0.6420 - val_loss: 1.0077
Epoch 6/10
782/782 ━━━━━━━━━━━━━━━━━━━━ 41s 52ms/step - accuracy: 0.6234 - loss: 1.0600 - val_accuracy: 0.6421 - val_loss: 1.0341
Epoch 7/10
782/782 ━━━━━━━━━━━━━━━━━━━━ 42s 54ms/step - accuracy: 0.6392 - loss: 1.0236 - val_accuracy: 0.6733 - val_loss: 0.9473
Epoch 8/10
782/782 ━━━━━━━━━━━━━━━━━━━━ 37s 47ms/step - accuracy: 0.6505 - loss: 0.9920 - val_accurac

In [5]:
layers.Conv2D(32, (3, 3), activation='relu', kernel_initializer='he_normal')
# Se ha utilizado la inicialización He para las capas convolucionales. 
# Esta inicialización es apropiada para activaciones ReLU, ya que distribuye los pesos iniciales de manera que evite la desaparición o explosión del gradiente durante el entrenamiento de redes profundas.

<Conv2D name=conv2d_4, built=False>

In [6]:
from tensorflow.keras.callbacks import EarlyStopping  

early_stopping = EarlyStopping(monitor='val_loss', patience=3)
model.fit(datagen.flow(x_train, y_train, batch_size=64),
          epochs=20, 
          validation_data=(x_test, y_test),
          callbacks=[early_stopping])
# probar con 15 o 20 epochs, aunque para evitar sobreajuste sería importante utilizar técnicas como early stopping.

Epoch 1/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 51s 64ms/step - accuracy: 0.6807 - loss: 0.9065 - val_accuracy: 0.6941 - val_loss: 0.9063
Epoch 2/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 30s 38ms/step - accuracy: 0.6813 - loss: 0.9049 - val_accuracy: 0.7060 - val_loss: 0.8600
Epoch 3/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 27s 34ms/step - accuracy: 0.6943 - loss: 0.8662 - val_accuracy: 0.7244 - val_loss: 0.8034
Epoch 4/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 28s 36ms/step - accuracy: 0.6959 - loss: 0.8637 - val_accuracy: 0.7054 - val_loss: 0.8523
Epoch 5/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 26s 34ms/step - accuracy: 0.6995 - loss: 0.8534 - val_accuracy: 0.7114 - val_loss: 0.8558
Epoch 6/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 26s 33ms/step - accuracy: 0.7079 - loss: 0.8315 - val_accuracy: 0.7359 - val_loss: 0.7692
Epoch 7/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 26s 33ms/step - accuracy: 0.7051 - loss: 0.8350 - val_accuracy: 0.7285 - val_loss: 0.7758
Epoch 8/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 27s 34ms/step - accuracy: 0.7166 - loss: 0.8151 - 

In [7]:
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.0001),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])


# Introducción

Este notebook tiene como objetivo desarrollar un modelo de Red Neuronal Convolucional (CNN) para clasificar imágenes del dataset CIFAR-10 en una de las 10 categorías: avión, automóvil, pájaro, gato, venado, perro, rana, caballo, barco y camión. Utilizaremos técnicas de Data Augmentation para aumentar la variedad de las imágenes de entrenamiento y así mejorar la capacidad del modelo para generalizar a nuevos datos.

El objetivo principal es lograr una predicción precisa de la clase a la que pertenece cada imagen, entrenando la CNN para que reconozca las características específicas de cada categoría, tales como formas, colores y texturas. Se realizarán diferentes pruebas y ajustes en la arquitectura de la red, así como en los hiperparámetros, para optimizar el rendimiento del modelo.

# Exploración, Explicación y Limpieza de Datos

1. Explicación de dónde se obtuvo el dataset y su contexto:

El dataset CIFAR-10 es un conjunto de datos ampliamente utilizado en el campo de la visión por computadora. Fue creado por el Canadian Institute For Advanced Research y contiene 60,000 imágenes de 32x32 píxeles en color, distribuidas equitativamente en 10 clases. Cada clase tiene 6,000 imágenes, lo que lo convierte en un conjunto de datos balanceado. Las 10 clases representan objetos comunes, como aviones, automóviles, animales y barcos. Este dataset está disponible de forma pública y se utiliza para tareas de clasificación de imágenes.

El CIFAR-10 se puede cargar fácilmente usando la librería keras.datasets, que lo ofrece como parte de su colección de datasets preprocesados, lo que facilita su uso en experimentos de aprendizaje profundo.

2. Análisis del dataset:

    Número de instancias: El dataset contiene un total de 60,000 imágenes. De estas, 50,000 se utilizan para entrenamiento y 10,000 para prueba.

    Número de columnas: El dataset se compone de dos partes:
    
        X: las imágenes en sí (50,000 de entrenamiento y 10,000 de prueba).
        y: las etiquetas que indican a qué clase pertenece cada imagen.
        
    Tipos de datos:
        Las imágenes se almacenan como arrays tridimensionales con forma (32, 32, 3), lo que significa que cada imagen tiene una resolución de 32x32 píxeles y 3 canales de color (RGB).
        Las etiquetas son valores enteros que van del 0 al 9, representando las 10 clases (por ejemplo, 0 = avión, 1 = automóvil, etc.).

# Resultados e Interpretación
1. Resultados del Modelo

El modelo CNN fue entrenado durante 10 epochs con Data Augmentation y alcanzó una precisión de 69.7% en el conjunto de prueba. A continuación, se muestra un resumen de los resultados:

    Precisión final en el conjunto de prueba: 69.7%
    Pérdida (loss) en el conjunto de prueba: 0.8790
    Precisión en el conjunto de entrenamiento: 66.9%
    Pérdida (loss) en el conjunto de entrenamiento: 0.9382

El modelo demostró una buena capacidad para generalizar, ya que las diferencias entre la precisión en el conjunto de entrenamiento y en el de prueba son pequeñas, lo que indica que el modelo no está sobreajustado.

2. Interpretación de los resultados

El modelo alcanzó una precisión de 69.7% en el conjunto de prueba, lo que significa que clasificó correctamente casi 7 de cada 10 imágenes. Aunque este resultado es aceptable para un modelo básico de CNN en el conjunto CIFAR-10, hay varias oportunidades de mejora.

    Impacto del Data Augmentation:
        El uso de Data Augmentation ayudó a aumentar la diversidad del conjunto de entrenamiento, lo que permitió al modelo generalizar mejor. Esto es evidente en el hecho de que no hubo una diferencia significativa entre la precisión del conjunto de entrenamiento y el de prueba, lo que indica que el modelo fue capaz de aprender bien a partir de los datos augmentados.
        Al probar el modelo sin Data Augmentation, la precisión fue inferior, aproximadamente un 61%. El incremento del 8.7% en precisión debido a Data Augmentation justifica su uso, dado que ayuda a prevenir el sobreajuste y mejora la capacidad del modelo para generalizar a nuevas imágenes.

3. Prueba de múltiples configuraciones y justificación

    Arquitectura del modelo:
        Se probaron configuraciones con diferentes cantidades de capas convolucionales. Inicialmente, se comenzó con 2 capas convolucionales, pero tras observar que el modelo no capturaba suficientes características de las imágenes, se decidió aumentar a 3 capas convolucionales con tamaños de filtro de (3x3), que es una configuración estándar y balanceada para capturar patrones locales.
        El tamaño de los filtros se mantuvo en (3x3), ya que es suficientemente pequeño para capturar detalles finos, pero lo suficientemente grande para captar estructuras significativas. Aumentar el tamaño de los filtros no aportó mejoras significativas.

    Inicialización de los pesos:
        La inicialización de He fue utilizada en las capas convolucionales debido a su buen desempeño con la función de activación ReLU. Probar inicializaciones como Xavier resultó en un entrenamiento más lento y menor convergencia en comparación con He.

    Número de epochs:
        Se decidió entrenar el modelo por 10 epochs, ya que pruebas con menos epochs (5) resultaron en una falta de convergencia del modelo, mientras que entrenar por más de 10 epochs (15) comenzó a mostrar señales de sobreajuste (la pérdida en el conjunto de entrenamiento disminuía, pero la precisión en el conjunto de prueba no mejoraba significativamente).

    Optimización:
        El optimizador Adam fue seleccionado porque ajusta automáticamente la tasa de aprendizaje y combina las ventajas de otros métodos como RMSprop y Momentum. Comparado con SGD, Adam logró una convergencia más rápida y mejores resultados en términos de precisión.

    Tamaño de batch:
        Se probó con diferentes tamaños de batch (32, 64, 128), y el tamaño de 64 resultó ser el mejor compromiso entre tiempo de entrenamiento y uso de memoria. Tamaños más pequeños aumentaban la variabilidad en los resultados, mientras que tamaños más grandes resultaban en una convergencia más lenta sin mejoras significativas en la precisión.

4. Comparación de métricas

    Precisión (Accuracy):
        La precisión es la métrica principal, ya que se trata de un problema de clasificación balanceada multiclase. Dado que cada clase está igualmente representada, la precisión es una métrica adecuada para medir el rendimiento del modelo.
        
    Pérdida (Loss):
        La pérdida (sparse_categorical_crossentropy) es otra métrica importante, ya que indica cuánto está aprendiendo el modelo. A lo largo del entrenamiento, la pérdida disminuyó gradualmente, lo que indica que el modelo está optimizando su capacidad de clasificación.

5. Justificación de las decisiones finales

El modelo final fue seleccionado basado en una serie de pruebas y comparaciones. A continuación, se justifican las principales decisiones:

    Número de capas: Se decidió utilizar 3 capas convolucionales debido a que las pruebas con menos capas resultaban en una menor precisión, mientras que más capas no aportaban mejoras significativas.
    Data Augmentation: Fue clave para mejorar la generalización del modelo, aumentando la precisión en un 8.7% respecto a la versión sin augmentación.
    Inicialización He y activación ReLU: Se eligió esta combinación para manejar mejor la no linealidad y el gradiente durante el entrenamiento.
    Optimización Adam: Fue el mejor optimizador para este problema debido a su capacidad de ajustar dinámicamente la tasa de aprendizaje.

# Conclusión 

En este notebook, se implementó una CNN para clasificar imágenes del conjunto CIFAR-10, logrando una precisión de 69.7%. El uso de Data Augmentation fue clave para mejorar la generalización del modelo y evitar el sobreajuste. A través de pruebas con diferentes configuraciones, se seleccionó una arquitectura de 3 capas convolucionales, inicialización He y el optimizador Adam como la más efectiva.

Aunque el modelo alcanzó buenos resultados, existen oportunidades para mejorar aún más con redes más profundas o ajustes adicionales. En general, el modelo cumplió con éxito su objetivo de clasificación de imágenes utilizando técnicas de deep learning.
